source/validate_layout.cpp

```diff
@@ -0,0 +1,205 @@
+
+// Source code for logical layout validation as described in section 2.4
+
+#include "validate_types.h"
+#include "validate_passes.h"
+#include "libspirv/libspirv.h"
+
+#include "diagnostic.h"
+#include "opcode.h"
+#include "operand.h"
+
+#include <cassert>
+
+using namespace libspirv;
+
+namespace {
+
+// Module scoped instructions are processed by determining if the opcode
+// is part of the current layout section. If it is not then the next sections is
+// checked.
+spv_result_t ModuleScopedInstructions(ValidationState_t& _,
+                                      const spv_parsed_instruction_t* inst,
+                                      SpvOp opcode) {
+  while (_.isOpcodeInCurrentLayoutSection(opcode) == false) {
+    _.progressToNextLayoutSectionOrder();
+
+    switch (_.getLayoutSection()) {
+      case kLayoutMemoryModel:
+        if (opcode != SpvOpMemoryModel) {
+          return _.diag(SPV_ERROR_INVALID_LAYOUT)
+                 << spvOpcodeString(opcode)
+                 << " cannot appear before the memory model instruction";
+        }
+        break;
+      case kLayoutFunctionDeclarations:
+        // All module sections have been processed. Recursivly call
+        // ModuleLayoutPass to process the next section of the module
+        return libspirv::ModuleLayoutPass(_, inst);
+      default:
+        break;
+    }
+  }
+  return SPV_SUCCESS;
+}
+
+// Function declaration validation is performed by making sure that the
+// FunctionParameter and FunctionEnd instructions only appear inside of
+// functions. It also ensures that the Function instruction does not appear
+// inside of another function. This stage ends when the first label is
+// encountered inside of a function.
+spv_result_t FunctionScopedInstructions(ValidationState_t& _,
+                                        const spv_parsed_instruction_t* inst,
+                                        SpvOp opcode) {
+  if (_.isOpcodeInCurrentLayoutSection(opcode)) {
+    switch (opcode) {
+      case SpvOpFunction:
+        if (_.in_function_body()) {
+          return _.diag(SPV_ERROR_INVALID_LAYOUT)
+                 << "Cannot declare a function in a function body";
+        }
+        spvCheckReturn(_.get_functions().RegisterFunction(
+            inst->result_id, inst->type_id,
+            inst->words[inst->operands[2].offset],
+            inst->words[inst->operands[3].offset]));
+        if (_.getLayoutSection() == kLayoutFunctionDefinitions)
+          spvCheckReturn(_.get_functions().RegisterSetFunctionDeclType(
+              FunctionDecl::kFunctionDeclDefinition));
+        break;
+
+      case SpvOpFunctionParameter:
+        if (_.in_function_body() == false) {
+          return _.diag(SPV_ERROR_INVALID_LAYOUT) << "Function parameter "
+                                                     "instructions must be in "
+                                                     "a function body";
+        }
+        if (_.get_functions().get_block_count() != 0) {
```

Comment:
When a label is encountered a new block is added to the function increasing the count to 1. The only instruction which can appear between the function instruction and the first label is the function parameter and the line instruction. I don't know if the line instruction is valid between the function and function parameter instructions.